### PR TITLE
Add kubos host script

### PR DESCRIPTION
**Pull request by kyleparrott**
- Rename kubos.py to kubos-sdk.py. 
  - Kubos-sdk.py is now only called inside the kubos-sdk docker container

- Add new File kubos.py as the "main" script called on the host machine
 - manages kubos-sdk container image
 - passes commands through to instances of kubos-sdk containers

### CODE REVIEW

File: kubos-sdk.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+
+def _init(name):
+    print 'Initializing project: %s ...' % name
+    c = component.Component(os.getcwd())
+    c.description['name'] = name
+    c.description['bin'] = './source'
+    c.description['dependencies'] = {}
+    c.description['homepage'] = 'https://<homepage>'
+    c.description['dependencies']["kubos-rt"] = "".join([org_name, '/', kubos_rt, '#', kubos_rt_branch])
```
Comment (marshall) [inline kubos-sdk.py]: nit: can you combine this with the initializer for `c.description['dependencies']` on line 47?

File: kubos-sdk.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    init.initNonInteractive(None, c)
+
+
+def _target(args):
```
Comment (marshall) [inline kubos-sdk.py]: nit: `args` should just be called `target` here, and I'm actually thinking this logic is simple enough to belong in `main()` rather than it's own separate function, but I'd be OK merging as is

File: kubos-sdk.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+def get_current_target():    
+    with open(yotta_meta_file, 'r') as meta_file:
+        data = json.load(meta_file)
+        target_str = str(data['build']['target'])
+        return target_str.split(',')[0]
+
+
+def cmd(*args, **kwargs):
+    # print ' '.join(args)
```
Comment (marshall) [inline kubos-sdk.py]: nit leftover debugging comment :)

File: kubos.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+	if args.update:
+		update()
+	elif unknown_args:
+		pass_through(*unknown_args)
+	else:
+		parser.print_help()
+		pass_through('--help')
+
+def update(cli):
+	print "updating"
```
Comment (marshall) [inline kubos.py]: It looks like you left `cli` as a function argument even though you ignore it and statically initialize with a call to `get_cli`. would be good to either remove the function argument or initialize the docker cli in `main` and pass it along

File: kubos.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+	cli.start(container_id, binds={
+		cwd : {
+			'bind': cwd,
+			'ro': False
+		}	
+	})
+
+	cli.stop(container_id)
+	container_output = cli.logs(container=container_id, stdout=True, stderr=True, stream=False)
```
Comment (marshall) [inline kubos.py]: I'm worried that we're not printing log data until the passed in command is finished executing. especially for building, won't the process just show nothing until the build process is finished?

I'm curious what happens when you pass in `cli.logs(.., stream=True)`. is it possible to print console output as we get it this way before calling `cli.stop()`?

**Comment by marshall** [comment]
@kyleparrott review complete, check comments :)

**Comment by kyleparrott** [comment]
@marshall ready for re-review

File: kubos-sdk.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+
+from yotta import init, target
+from yotta.lib import component, globalconf
+
+kubos_rt = 'kubos-rt'
+kubos_rt_branch = 'master'
+org_name = 'openkosmosorg'
+kubos_rt_full_path = '%s@%s/%s#%s' % (kubos_rt, org_name, kubos_rt, kubos_rt_branch)
+
+KubOS_manifest_url = 'https://raw.githubusercontent.com/openkosmosorg/kubos-manifest/master/default.xml'
```
Comment (marshall) [inline kubos-sdk.py]: nit: looks like this variable is no longer needed

File: kubos-sdk.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+
+yotta_meta_file = '.yotta.json'
+yotta_install_path = '/usr/local/bin/yotta'
+
+def main():
+    parser = argparse.ArgumentParser(description = 'Kubos SDK')
+    parser.add_argument('--init', nargs='?', type=str, help='Create a new module')
+    parser.add_argument('--target', nargs='?', type=str, const='_show_current_target_', help='Set target device')
```
Comment (marshall) [inline kubos-sdk.py]: Can you make these sub-commands instead of arguments to match yotta? i.e.:

```
$ kubos init <name>
$ kubos target (<target>)
```

File: kubos-sdk.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+kubos_rt_full_path = '%s@%s/%s#%s' % (kubos_rt, org_name, kubos_rt, kubos_rt_branch)
+
+KubOS_manifest_url = 'https://raw.githubusercontent.com/openkosmosorg/kubos-manifest/master/default.xml'
+
+yotta_meta_file = '.yotta.json'
+yotta_install_path = '/usr/local/bin/yotta'
+
+def main():
+    parser = argparse.ArgumentParser(description = 'Kubos SDK')
+    parser.add_argument('--init', nargs='?', type=str, help='Create a new module')
```
Comment (marshall) [inline kubos-sdk.py]: nit: based on the logic below, I think the module name should actually required here, but you're saying it's optional with the `nargs='?'` option. maybe change to `nargs=1`, and verify it throws an error?

File: kubos-sdk.py [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+    args, anonymous_args = parser.parse_known_args()
+    globalconf.set('interactive', False)
+    
+    if args.init:
+        _init(args.init)
+    elif args.target:
+        if args.target == '_show_current_target_':
+            show_target()
+        else:
+            set_target(args)
```
Comment (marshall) [inline kubos-sdk.py]: shouldn't this be `set_target(args.target)`?

File: kubos-sdk.py [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+
+def set_target(new_target):
+    print 'Setting Target: %s' % new_target.split('@')[0]
+    globalconf.set('plain', False)
+    target_args = argparse.Namespace(set_target=new_target, save_global = False, no_install = False)
+    target.execCommand(target_args, '') 
+
+
+def get_current_target():    
+    with open(yotta_meta_file, 'r') as meta_file:
```
Comment (marshall) [inline kubos-sdk.py]: what happens here if `kubos target` is called before the yotta meta file exists? do you know if the meta file is created at installation time?

Comment (marshall) [inline kubos-sdk.py]: also, `yotta_meta_file` should be defined as an absolute path, I'm assuming it's supposed to be in the home directory?

File: kubos.py [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+import sys
+
+from docker import Client
+import docker
+
+container_name = 'kubostech/kubos-sdk'
+
+def main():
+	parser = argparse.ArgumentParser('kubos-host script for Kubos SDK')
+	parser.add_argument('--update', action='store_true', help='pull latest sdk container')
```
Comment (marshall) [inline kubos.py]: let's also make update a sub-command (see my comment on `kubos-sdk.py`)

File: kubos.py [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+	if args.update:
+		update()
+	elif unknown_args:
+		pass_through(*unknown_args)
+	else:
+		parser.print_help()
+		pass_through('--help')
+
+def update():
+	print "updating"
```
Comment (marshall) [inline kubos.py]: nit: either make this message a bit more informative, or remove it :)

File: kubos.py [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+def pass_through(*args):
+	cwd = os.getcwd() 
+	cli = get_cli()
+	python = '/usr/bin/python'
+	sdk_script = '/kubos-sdk/kubos-sdk.py' 
+	kubos_sdk_cmd =  ' '.join([python, sdk_script])
+	arg_list = list(args)
+	arg_list.insert(0, kubos_sdk_cmd)
+
+	command = ' '.join(arg_list)
```
Comment (marshall) [inline kubos.py]: This is probably bad -- joining arguments with a space without escaping like this can cause problems for any arguments (like directory/file names) that potentially have a space in them.

I looked over the [Docker Client API docs for `create_container`](http://docker-py.readthedocs.io/en/latest/api/?highlight=create_container#create_container) and it looks like `command` can either be a string or a list, which implies that the API will safely escape the list of arguments for us when it's passed onto the container.

**tl;dr** Let's pass `arg_list` directly here instead of trying to construct our own string

**Comment by kyleparrott** [comment]
@marshall, ready for another pass

File: kubos-sdk.py [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+        target.displayCurrentTarget(target_args)
+    else:
+        print 'No target currently set'
+
+
+
+def set_target(new_target):
+    print 'Setting Target: %s' % new_target.split('@')[0]
+    globalconf.set('plain', False)
+    target_args = argparse.Namespace(set_target=new_target, save_global = False, no_install = False)
```
Comment (marshall) [inline kubos-sdk.py]: nit: keep consistent spacing on your kwargs here :)

File: kubos-sdk.py [hunk 15]
```diff
@@ -0,0 +1,103 @@
+#!/usr/bin/env python
```
Comment (marshall) [inline kubos-sdk.py]: nit: add Apache 2.0 copyright header

File: kubos-sdk.py [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+import json
+import os
+import subprocess
+import sys
+import urllib2 
+import xml.etree.ElementTree as ET
+
+from yotta import init, target
+from yotta.lib import component, globalconf
+
```
Comment (marshall) [inline kubos-sdk.py]: nit: can you add a comment here describing that this is the 'target' script that usually runs in Docker?

File: kubos.py [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+		pass_through(command, proj_name)
+	elif command == 'target':
+		target = provided_args['target'] 
+		if target:
+			pass_through(command, target)
+		else:
+			pass_through(command)
+	elif command == 'update':
+		update()
+
```
Comment (marshall) [inline kubos.py]: Can you print a help message if they give an unknown command?

**Comment by marshall** [comment]
@kyleparrott review complete, go ahead and fix up the nits and add the help message, and feel free to merge

File: kubos.py (hunk 17 above)
Comment (kyleparrott) [inline kubos.py]: Argparse will throw an error already with a default help message. See screenshot for example. I can still investigate to put a custom message if you would like. 

screen shot 2016-05-20 at 11 00 37 am

Comment (marshall) [inline kubos.py]: nope, no need for now. argparse's error message looks just fine!
